executor/cook/subprocess.py

```diff
@@ -0,0 +1,216 @@
+
+    Returns
+    -------
+    whether the process is still running.
+    """
+    return process.poll() is None
+
+
+def find_process_group(process_id):
+    """Return the process group id of the process with process id process_id.
+    Parameters
+    ----------
+    process_id: int
+        The process id.
+    Returns
+    -------
+    The process group id of the process with process id process_id or None.
+    """
+    try:
+        group_id = os.getpgid(process_id)
+        logging.info('Process (pid: {}) belongs to group (id: {})'.format(process_id, group_id))
+        return group_id
+    except ProcessLookupError:
+        logging.info('Unable to find group for process (pid: {})'.format(process_id))
+    except Exception:
+        logging.exception('Error in finding group for process (pid: {})'.format(process_id))
+
+
+def _send_signal_to_process(process_id, signal_to_send):
+    """Send the signal_to_send signal to the process with process_id.
+    Parameters
+    ----------
+    process_id: int
+        The id of the process whose group to kill.
+    signal_to_send: signal.Signals enum
+        The signal to send to the process group.
+    Returns
+    -------
+    True if the signal was sent successfully.
+    """
+    signal_name = signal_to_send.name
+    try:
+        logging.info('Sending signal {} to process (id: {})'.format(signal_name, process_id))
+        os.kill(process_id, signal_to_send)
+        return True
+    except ProcessLookupError:
+        logging.info('Unable to send signal {} as could not find process (id: {})'.format(signal_name, process_id))
+    except Exception:
+        logging.exception('Error in sending signal {} to process (id: {})'.format(signal_name, process_id))
+    return False
+
+
+def _send_signal_to_process_tree(process_id, signal_to_send):
+    """Send the signal_to_send signal to the process tree rooted at process_id.
+    Parameters
+    ----------
+    process_id: int
+        The id of the root process.
+    signal_to_send: signal.Signals enum
+        The signal to send to the process group.
+    Returns
+    -------
+    True if the signal was sent successfully.
+    """
+    signal_name = signal_to_send.name
+    try:
+        logging.info('Sending signal {} to process tree rooted at (id: {})'.format(signal_name, process_id))
+        process = psutil.Process(process_id)
+        children = process.children(recursive=True)
+
+        _send_signal_to_process(process_id, signal_to_send)
+
+        signal_sent_to_all_children_successfully = True
+        if children:
+            for child in children:
+                logging.info('Found child process (id: {}) of process (id: {})'.format(child.pid, process_id))
+                if not _send_signal_to_process(child.pid, signal_to_send):
+                    logging.info('Failed to send signal {} to child process (id: {})'.format(signal_name, child.pid))
+                    signal_sent_to_all_children_successfully = False
```

Comment:
Did we look into whether this does the kills in the same order as the Mesos command executor?

Reply:
I haven't yet.

Reply:
I was previously using depth-first order whereas [stout](https://github.com/deric/libprocess/blob/master/3rdparty/stout/include/stout/os/killtree.hpp#L123) (the library Mesos uses) used breadth-first order. I have changed the code to match the order and also use the `SIGSTOP` and `SIGCONT` signals.